OpenChange/MAPIStoreMailMessage.m

```diff
@@ -396,36 +412,99 @@ - (void) _fetchBodyData
             encoding = @"7-bit";
 
           /* We should provide a case for each of the types in acceptedMimeTypes */
-          if ([mimeType isEqualToString: @"text/html"] && !mailIsEvent)
-            {
-              content = [content bodyDataFromEncoding: encoding];
-              [htmlContent appendData: content];
-            }
-          else if ([mimeType isEqualToString: @"text/plain"] && !mailIsEvent)
+          if (!mailIsEvent)
             {
+              NSString *charset;
+              NSStringEncoding stringEncoding;
+              NSString *stringValue; 
+              BOOL html;
+              if ([mimeType isEqualToString: @"text/html"])
+                {
+                  html = YES;
+                  
+                }
+              else if ([mimeType isEqualToString: @"text/plain"])
+                {
+                  html = NO;
+                }
+              else
+                {
+                  [self warnWithFormat: @"Unsupported MIME type for non-event body part: %@.",
+                        mimeType];
+                  continue;
+                }
+                          
               content = [content bodyDataFromEncoding: encoding];
-              NSString *charset = [bodyPartsCharsets objectForKey: key];
+              charset = [bodyPartsCharsets objectForKey: key];
+              stringEncoding = [NSString stringEncodingForEncodingNamed: headerCharset];
```

Comment:
this variable name is quite misleading, imho it should be called `headerEncoding` and it can be set outside the loop because `headerCharset` will always have the same value.